### PR TITLE
Streamline WGR API to simplify usage; Standardize labels for level-0 reduction in binary WGR; Make addition of intercept to covariates consistent

### PR DESCRIPTION
**Pull request by kianfar77**
## What changes are proposed in this pull request?
This PR involves the following:
- It moves argument passing to WGR API to class level to avoid repeat passing of argument to each function. `RidgeReduction` class is used to encapsulate/generate all that is required for WGR. `RidgeRegression` and `LogisticRidgeRegression` can now be initialized based on a `RidgeReduction` as well. `set` and `get` functions are introduced for the classes for user to feed previously generated items to them if needed. A `estimate_loco_offsets` function is added that can perform the whole reduction and estimation in a single call. 
- The API now has the default option of automatically detecting whether `label_df` is binary or quantitative and performs the corresponding WGR. The default detection can be overriden by the `label_type` argument. 
- To match the REGENIE algorithm, the binary labels are now standardized in the first round of reduction (in `RidgeReduction`). 
-  To match REGENIE and avoid confusions, the addition of intercept is now the default behavior in all classes and can be turned off with the new `add_intercept` option. 
- The `linear_model` module is now removed and everything is moved to `wgr` module with some clarifying name changes.

## How is this patch tested?
- [x] Unit tests
- [ ] Integration tests
- [ ] Manual tests

(Details)

### CODE REVIEW

File: python/glow/wgr/linear_model/ridge_model.py [hunk 1]
```diff
@@ -19,153 +19,212 @@
...
+        state = self.__dict__.copy()
+        # Remove the unpicklable entries.
+        del state['_RidgeReduction__block_df'], state['_RidgeReduction__model_df'], state[
+            '_RidgeReduction__reduced_block_df']
+        return state
+
+    def __setstate__(self, state):
+        # Restore instance attributes
+        self.__dict__.update(state)
+        self.__block_df = [x for x in globals().values() if id(x) == self.__block_df_id]
```
[the remark this answers is not in the document]
Comment (henrydavidge) [reply python/glow/wgr/linear_model/ridge_model.py]: Actually, why do we even need this `__setstate__`? When is it called on the driver? I'm not sure we need these ids to be picklable.

**Review by henrydavidge — COMMENTED** [review]
I like the high level idea but have a number of comments on the implementation

File: python/glow/wgr/linear_model/functions.py [hunk 2]
```diff
@@ -521,68 +503,70 @@ def __check_binary(df: pd.DataFrame) -> None:
...
-        __check_standardized(labeldf, 'label')
+
+@typechecked
+def prepare_labels_and_warn(labeldf: pd.DataFrame, binary: bool, label_type: str) -> pd.DataFrame:
+    if label_type == 'detect':
+        if binary:
+            warnings.warn(
+                "The label DataFrame is binary. Ridge reduction for binary phenotypes will be applied.",
+                UserWarning)
+            return labeldf - labeldf.mean()
```
[the remark this answers is not in the document]
Comment (kianfar77) [reply python/glow/wgr/linear_model/functions.py:529]: No, here is where one of the differences with regenie is fixed. regenie only mean centers in level0 reduction for binary.

File: python/glow/wgr/linear_model/ridge_model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kianfar77) [reply python/glow/wgr/linear_model/ridge_model.py]: This is somewhat based on what you gave me (https://stackoverflow.com/questions/2345944/exclude-objects-field-from-pickling-in-python). As far as I undertand, the essence is to keep a picklable surrogate for the object, drop object from pickling in `__getstate__` and then revive the object inside the class in ``__setstate__`` using the surrogate. Here I am using the object id as the surrogate because I am dealing with a variable. In the example in the link, since it is dealing with a file, it is using filename. On my local tests this works and without it spark dfs do not pickle. Do you think there may be a problem with this on a real cluster?

File: python/glow/wgr/linear_model/ridge_model.py [hunk 3]
```diff
@@ -19,153 +19,212 @@
...
+
+    def set_model_df(self, model_df: DataFrame = None) -> None:
+        self.__model_df = model_df
+        self.__model_df_id = id(self.__model_df)
+
+    def set_reduced_block_df(self, reduced_block: DataFrame = None) -> None:
+        self.__reduced_block_df = reduced_block
+        self.__reduced_block_df_id = id(self.__reduced_block_df)
+
+    def is_binary(self) -> bool:
```
[the remark this answers is not in the document]
Comment (kianfar77) [reply python/glow/wgr/linear_model/ridge_model.py]: I wanted these to be exposed to the end user. Made it this way mostly to have a uniform naming pattern for these functions that lends itself to the case when we want to make changes to them that involve are complicated than just access. If you prefer, I can change them to `self.is_binary` etc (without `__`)

File: python/glow/wgr/linear_model/functions.py [hunk 4]
```diff
@@ -521,68 +503,70 @@ def __check_binary(df: pd.DataFrame) -> None:
...
+@typechecked
+def prepare_covariates(covdf: pd.DataFrame) -> pd.DataFrame:
+    return __fillna_and_standardize(covdf)
+
+
+@typechecked()
+def check_model(modeldf):
+    if modeldf is None:
+        raise ValueError(
+            'No model DataFrame found! Run fit() or provide a previously made model using set_model_df()'
```
[the remark this answers is not in the document]
Comment (kianfar77) [reply python/glow/wgr/linear_model/functions.py:559]: Yes

File: python/glow/wgr/linear_model/ridge_model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kianfar77) [reply python/glow/wgr/linear_model/ridge_model.py]: You are right. removed variable recovery from setstate

File: python/glow/wgr/linear_model/ridge_model.py [hunk 5]
```diff
@@ -178,140 +237,147 @@ class RidgeRegression:
     coefficients.  The optimal ridge alpha value is chosen for each label by maximizing the average out of fold r2
     score.
     """
-    def __init__(self, alphas: NDArray[(Any, ), Float] = np.array([])) -> None:
+    def __init__(
+        self, ridge_reduced: RidgeReduction, alphas: NDArray[(Any, ),
```
[the remark this answers is not in the document]
Comment (kianfar77) [reply python/glow/wgr/linear_model/ridge_model.py]: Done

File: python/glow/wgr/linear_model/ridge_model.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kianfar77) [reply python/glow/wgr/linear_model/ridge_model.py]: Things changed here a bit in this new push. Anything that has get and set here now is not simple access and I want it exposed to the user.

File: python/glow/wgr/linear_model/functions.py [hunk 6]
```diff
@@ -700,3 +684,16 @@ def apply_model_df(blockdf, modeldf, cvdf, transform_udf, transform_key_pattern,
         .groupBy(transform_key_pattern) \
         .apply(transform_udf) \
         .join(cvdf, ['label', 'alpha'], 'inner')
+
+
+def wgr(blockdf: DataFrame,
```
[the remark this answers is not in the document]
Comment (kianfar77) [reply python/glow/wgr/linear_model/functions.py]: Done

File: python/glow/wgr/linear_model/functions.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kianfar77) [reply python/glow/wgr/linear_model/functions.py]: Yes! remnants of my comparisons with regenie as it was easier to turn off scaling in regenie. Fixed

File: python/glow/wgr/wgr_functions.py [hunk 7]
```diff
@@ -199,3 +204,46 @@ def transpose_one(contig):
...
+    Args:
+        block_df : Spark DataFrame representing the beginning block matrix X
+        label_df : Pandas DataFrame containing the target labels used in fitting the ridge models
+        sample_blocks : Dict containing a mapping of sample_block ID to a list of corresponding sample IDs
+        cov_df : Pandas DataFrame containing covariates to be included in every model in the stacking
+            ensemble (optional).
+        add_intercept: If True, an intercept column (all ones) will be added to the covariates
+            (as the first column)
+        alphas : array_like of alpha values used in the ridge reduction (optional).
+        label_type: String to determine type treatment of labels. It can be 'detect' (default), 'binary',
```
Comment (henrydavidge) [inline python/glow/wgr/wgr_functions.py:231]: nit: Say what happens when the type is 'detect'

File: python/glow/wgr/wgr_functions.py [hunk 8]
```diff
@@ -199,3 +204,46 @@ def transpose_one(contig):
...
+
+    Args:
+        block_df : Spark DataFrame representing the beginning block matrix X
+        label_df : Pandas DataFrame containing the target labels used in fitting the ridge models
+        sample_blocks : Dict containing a mapping of sample_block ID to a list of corresponding sample IDs
+        cov_df : Pandas DataFrame containing covariates to be included in every model in the stacking
+            ensemble (optional).
+        add_intercept: If True, an intercept column (all ones) will be added to the covariates
+            (as the first column)
+        alphas : array_like of alpha values used in the ridge reduction (optional).
```
Comment (henrydavidge) [inline python/glow/wgr/wgr_functions.py:230]: nit: what happens if they're omitted?

File: python/glow/wgr/logistic_ridge_regression.py [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+        return obj
+
+    def __getstate__(self):
+        # Copy the object's state from self.__dict__ which contains
+        state = self.__dict__.copy()
+        # Remove the unpicklable entries.
+        del state['reduced_block_df'], state['model_df'], state['cv_df']
+        return state
+
+    def __setstate__(self, state):
```
Comment (henrydavidge) [inline python/glow/wgr/logistic_ridge_regression.py:97]: Do we actually need a custom `__setstate__` implementation? What happens if this is omitted?

File: python/glow/wgr/ridge_reduction.py [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+        self.reduced_block_df = None
+
+    def __getstate__(self):
+        # Copy the object's state from self.__dict__ which contains
+        state = self.__dict__.copy()
+        # Remove the unpicklable entries.
+        del state['block_df'], state['model_df'], state['reduced_block_df']
+        return state
+
+    def __setstate__(self, state):
```
Comment (henrydavidge) [inline python/glow/wgr/ridge_reduction.py:79]: Same question -- do we need a custom `__setstate__`?

File: python/glow/wgr/model_functions.py [hunk 11]
```diff
@@ -471,117 +474,168 @@ def __assert_all_present(df: pd.DataFrame, col_name: Any, df_name: str) -> None:
...
-        if num_non_binary_values == 0:
-            # Valid binary trait
-            continue
-
-        # Check as continuous trait
-        __assert_all_present(df, label, 'label')
-        if not continuous_ok:
+    if label_type == 'detect':
+        if is_binary:
+            warnings.warn(
```
Comment (henrydavidge) [inline python/glow/wgr/model_functions.py:557]: Should this really be a warning? This is expected. Seems like we should log or print.

**Review by henrydavidge — COMMENTED** [review]
Only minor comments on the main code. It looks great!

File: python/glow/wgr/wgr_functions.py [hunk 12]
```diff
@@ -199,3 +204,46 @@ def transpose_one(contig):
...
+                chromosomes will be inferred from the block matrix.
+
+    Returns:
+        Pandas DataFrame containing offset values per chromosome. The rows are indexed by sample ID and
+        chromosome; the columns are indexed by label. The column types are float64. The DataFrame is sorted using
+        chromosome as the primary sort key, and sample ID as the secondary sort key.
+    """
+    ridge_reduced = RidgeReduction(block_df, label_df, sample_blocks, cov_df, add_intercept, alphas,
+                                   label_type)
+    ridge_reduced.fit_transform()
```
Comment (henrydavidge) [inline python/glow/wgr/wgr_functions.py:243]: We probably want to cache this so it's not computed twice. We cache in our examples.

File: python/glow/wgr/logistic_ridge_regression.py [hunk 13]
```diff
@@ -0,0 +1,358 @@
...
+            loco_model_df = self.model_df.filter(
+                ~f.col('header').rlike(f'^chr_{chromosome}_(alpha|block)'))
+            self.model_df = loco_model_df
+            loco_y_hat_df = self.transform(response)
+            loco_y_hat_df['contigName'] = chromosome
+            y_hat_df = y_hat_df.append(loco_y_hat_df)
+            self.model_df = orig_model_df
+
+        self.y_hat_df = y_hat_df.set_index('contigName', append=True)
+        return self.y_hat_df
```
Comment (henrydavidge) [inline python/glow/wgr/logistic_ridge_regression.py:319]: Actually I think it would be appropriate to uncache the model dfs and cv dfs here since we've already materialized the offsets.

**Review by henrydavidge — APPROVED** [review]
I have a few minor comments outstanding. Please merge once that's in! Thanks Karen!!

File: python/glow/wgr/wgr_functions.py [hunk 14]
```diff
@@ -14,10 +14,15 @@
...
 from pyspark.sql import DataFrame, Row, SparkSession, SQLContext
 from typeguard import check_argument_types, check_return_type
-from typing import Dict, List
+from typing import Any, Dict, List
+from nptyping import Float, NDArray
+from .ridge_reduction import RidgeReduction
+from .ridge_regression import RidgeRegression
+from .logistic_ridge_regression import LogisticRidgeRegression
 
 __all__ = ['get_sample_ids', 'block_variants_and_samples', 'reshape_for_gwas']
```
Comment (henrydavidge) [inline python/glow/wgr/wgr_functions.py:27]: You need to add `estimate_loco_offsets` here.
